hibernate-reactive-core/src/main/java/org/hibernate/reactive/pool/impl/ProxyConnection.java

```diff
@@ -34,22 +42,82 @@ public ProxyConnection(ReactiveConnectionPool sqlClientPool, String tenantId) {
+			// in that case.
+			// There's also no guarantee that all consumers of 'connection' have finished when
+			// 'close()' is being called.
+			//
+			// 		Thread A						Thread B
+			//			withConnection():
+			//			read 'connection' field
+			//
+			//			(thread suspended)				close() runs and finishes
+			//
+			//			runs operation.apply(conn)
+			//
+			return operation.apply(conn);
 		}
-		else {
-			if ( connection == null ) {
-				// we're already in the process of fetching a connection,
-				// so this must be an illegal concurrent call
-				throw new IllegalStateException( "session is currently connecting to database" );
+
+		// No connection in the 'connection' field present.
+		//
+		// This can mean:
+		// 1. this ProxyConnection has been closed (duh)
+		// 2. no connection has been acquired from 'sqlClientPool'
+		// 3. a connection is currently being acquired from 'sqlClientPool'
+
+		// Tricky part comes next...
+		// Either no connection has been acquired from the 'sqlClientPool' yet or a connection
+		// is currently being acquired. Need to distinguish both cases and prevent that more than
+		// one connection's being acquired from 'sqlClientPool'.
+		//
+		// Generally, the CompletionStage returned from 'sqlClientPool.getConnection()' is used
+		// to "bundle" all outstanding 'operation's against the not-yet-available connection.
+		//
+		// Note: the 'connectionCompletionStage' is never cleared to 'null', because otherwise
+		// a new race condition might be introduced. However, it is set to null, in 'close()',
+		// because that's also guarded via 'stateChange'.
+		//
+		CompletionStage<ReactiveConnection> completionStage;
+		synchronized (stateChange) {
+			if (closed) {
+				// easy case, just bark and bite
+				throw new IllegalStateException("Proxy connection already closed");
+			}
+
+			completionStage = connectionCompletionStage;
+			if (completionStage == null) {
+				// First one getting here for this ProxyConnection instance.
+				// Start acquiring a connection from 'sqlClientPool' and store the
+				// CompletionStage instance for future use.
+				CompletionStage<ReactiveConnection> connection =
+						tenantId == null ? sqlClientPool.getConnection() : sqlClientPool.getConnection(tenantId);
+				completionStage = connection.thenApply(newConnection -> {
+					// Again, change to 'connection' guarded via 'stateChange'.
+					synchronized (stateChange) {
+						this.connection = newConnection;
+					}
+					return newConnection;
+				});
+				connectionCompletionStage = completionStage;
 			}
-			return operation.apply( connection );
 		}
+
+		// "Queue up" the 'operation'.
+		// If the connection's already available, the 'operation' will likely be executed
+		// via this call site.
+		// If the connection's not yet available, the 'operation' will be executed once
+		// the 'completionStage' to acquire the connection has finished.
+		return completionStage.thenCompose(operation);
```

Comment:
+1 I agree we need to compose subsequent calls into the "being acquired" handle here.

But I'm not sure that synchronization and volatiles are needed here - do you have more details about that or are you just adding them to be one the safe side?

Reply:
Are you really sure. Read the comment in the deleted code. Clearly I thought about this and decided the only way it could be null is in the case of a concurrent call, which it still seems to me is clearly wrong.

I'm still not convinced there is a bug here.

Reply:
> not sure that synchronization and volatiles are needed here

Yup, those are needed to prevent races. You cannot guarantee that all invocations there will happen on the same thread. So visibility-issues come into play - `volatile` mitigates some of those. x86 platforms are rather "friendly" wrt load/store instructions reordering to people, but other platforms like arm reorder instructions quite aggressively - so being very careful is not bad per se. If there's a chance that one field is accessed concurrently, `volatile` is your friend. I've seen such reordering issues in the past and those are really hard to catch - up to the point that other threads never saw a specific change.

Reply:
No I'm not really sure - I just did a session in pair with @DavideD trying to create a different test (w/o using threads) and we couldn't trigger the issue, however I didn't get to emulate a limited pool yet.

But if the state is possible, composing them seems fair. But +1 to understand more how this could happen.